lib/auth/saml.go

```diff
@@ -158,11 +200,28 @@ func (a *Server) calculateSAMLUser(connector types.SAMLConnector, assertionInfo
 
 	p.traits = services.SAMLAssertionsToTraits(assertionInfo)
 
+	diagInfo := func(infoType types.SSOInfoType, value interface{}) {
+		a.ssoDiagInfo(ctx, types.KindSAML, request.ID, infoType, value)
+	}
+
+	diagInfo(types.SSOInfoType_SAML_TRAITS_FROM_ASSERTIONS, p.traits)
+	diagInfo(types.SSOInfoType_SAML_CONNECTOR_TRAIT_MAPPING, connector.GetTraitMappings())
```

Comment:
These "diagInfo" calls look a bit out of place throughout the function now, I would consider refactoring/organizing it a bit like this:

- Introduce a concept of the "login context" - a struct that will be keeping the context for this login flow.
- Propagate login context to each method that needs it and will be attaching diagnostic info to it.
- Once the login is done (fail or success), save the diag info in the backend.
- I will also reiterate that I'm not a fan of using int "type" + interface, I would much prefer a bunch of typed structs even if it's a little more verbose. Or just use plain JSON.

So the flow would be similar to what happens here already, just a bit more structured IMO.

Reply:
Think of the `diagInfo` calls as similar to logging. This is the same pattern really:
1. We want to extract some information about stuff that is happening during execution.
2. It doesn't change the control flow in any way. We should be able to replace it with no-op and still be good.
3. It happens in potentially unrelated parts of the system.

I used explicit passing of context in the past (https://github.com/gravitational/teleport/pull/10393) and it wasn't great:
1. It felt like unnecessary complexity.
2. Passing the service boundary was a pain.
3. There were ways to get the control flow interrupted in a way that resulted in the context being simply lost and unavailable to the caller.
4. We were hurting backward compatibility by changing the APIs quite a lot.

Reply:
On further thought and rereading the suggestion, I think that perhaps some middle ground can be achieved. Right now all of the action is happening in `saml.go`, so maybe there is some room for refactoring here. For sure we could reduce the number of calls to the backend and drop the range iteration on retrieval, which is something.